### PR TITLE
[issue 18] Create basic interfaces

### PR DESCRIPTION
**Pull request by Serhiiilnytskyi**
## Issue #18

Changes:
1. Created base project structure.
2. Added annotation @Component
3. Added interfaces: ApplicationContext, BeanFactory, BeanPostProcessor, BeanDefinitionReader, ConfigBeanDefinitionReader, BeanDefinition, BeanDefinitionRegistry.
4. Covered new interfaces by Java documentation.
5. Added Intellij IDEA folder to gitignore.

## Checklist before requesting a review
- [ ] There are no MR remarks
- [x] Clean and concise code that is easy to read and understand
- [ ] Code covered with tests and they are green
- [x] Code API has informative JavaDoc on it
- [ ] Rich logging with detailed messages and correct use of log levels

### CODE REVIEW

File: src/main/java/com/bringframework/annotations/Component.java [hunk 1]
```diff
@@ -0,0 +1,23 @@
...
+@Retention(RUNTIME)
+public @interface Component {
+
+  /**
+   * The name of the bean. If no value is specified, the name of the class is used.
+   *
+   * @return the name of the bean
+   */
+  String value() default "";
+}
```
Comment (vitshch) [inline src/main/java/com/bringframework/annotations/Component.java]: GitHub shows warning that this file don't have new line at the end

File: src/main/java/com/bringframework/annotations/Component.java [hunk 2]
```diff
@@ -0,0 +1,23 @@
...
+
+import java.lang.annotation.Retention;
+import java.lang.annotation.Target;
+
+/**
+ * Annotation used to mark a class as a bean that can be managed by an IoC container. The class will be instantiated and
+ * configured by the container, and its lifecycle will be managed.
+ */
+@Target({TYPE})
+@Retention(RUNTIME)
```
Comment (vitshch) [inline src/main/java/com/bringframework/annotations/Component.java:16]: Add `@Documented` annotation

File: src/main/java/com/bringframework/factory/BeanFactory.java [hunk 3]
```diff
@@ -0,0 +1,17 @@
...
+ */
+public interface BeanFactory {
+
+  /**
+   * Creates and initializes beans from a collection of {@link BeanDefinition}s.
+   *
+   * @param beanDefinitions the collection of {@link BeanDefinition}s to create and initialize beans from
+   */
+  void createBeans(Collection<BeanDefinition> beanDefinitions);
+}
```
Comment (vitshch) [inline src/main/java/com/bringframework/factory/BeanFactory.java]: GitHub shows warning that this file don't have new line at the end

**Review by vitshch — CHANGES_REQUESTED** [review]

File: src/main/java/com/bringframework/annotations/Component.java (hunk 2 above)
Comment (Serhiiilnytskyi) [reply src/main/java/com/bringframework/annotations/Component.java:16]: Added.

Comment (Serhiiilnytskyi) [reply src/main/java/com/bringframework/annotations/Component.java:16]: Fixed.

File: src/main/java/com/bringframework/factory/BeanFactory.java (hunk 3 above)
Comment (Serhiiilnytskyi) [reply src/main/java/com/bringframework/factory/BeanFactory.java]: Fixed

File: src/main/java/com/bringframework/annotations/Component.java (hunk 1 above)
Comment (Serhiiilnytskyi) [reply src/main/java/com/bringframework/annotations/Component.java]: Fixed

**Pull request closed without merging**